### ISSUE TITLE
NodePort service is not reachable outside cluster using extraPortMappings

### ISSUE THREAD

**Issue by gagara** [kubernetes-sigs/kind#808]
**What happened**:
NodePort service created in cluster is not reachable from outside cluster when using extraPortMappings config.
 
**What you expected to happen**:
Service should be accessible by configured port. 

**How to reproduce it (as minimally and precisely as possible)**:
1. create kind cluster with extraPortMappings configuration
2. inside cluster create NodePort service with same port as in step 1.
3. try to call service from host by given port.

**Anything else we need to know?**:
kind config:
```
kind: Cluster
apiVersion: kind.sigs.k8s.io/v1alpha3
nodes:
- role: control-plane
  extraPortMappings:
  - containerPort: 8080
    hostPort: 8080
```
Detailed steps:
```
$ kind version
v0.5.0

$ kind create cluster --config .kube/kind-config.yaml 
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.15.3) 🖼
 ✓ Preparing nodes 📦 
 ✓ Creating kubeadm config 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Cluster creation complete. You can now use the cluster with:

export KUBECONFIG="$(kind get kubeconfig-path --name="kind")"
kubectl cluster-info

$ export KUBECONFIG="$(kind get kubeconfig-path --name="kind")"

$ docker ps
CONTAINER ID        IMAGE                  COMMAND                  CREATED              STATUS              PORTS                                                          NAMES
b1ba266386d7        kindest/node:v1.15.3   "/usr/local/bin/entr…"   About a minute ago   Up About a minute   0.0.0.0:8080->8080/tcp, 44605/tcp, 127.0.0.1:44605->6443/tcp   kind-control-plane

$ kubectl create deployment hello-node --image=gcr.io/hello-minikube-zero-install/hello-node
deployment.apps/hello-node created

$ kubectl expose deployment hello-node --type=NodePort --port=8080
service/hello-node exposed

$ kubectl get pods
NAME                          READY   STATUS    RESTARTS   AGE
hello-node-55b49fb9f8-g59kw   1/1     Running   0          112s

$ kubectl get svc
NAME         TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)          AGE
hello-node   NodePort    10.96.230.243   <none>        8080:30138/TCP   102s
kubernetes   ClusterIP   10.96.0.1       <none>        443/TCP          3m9s

$ docker exec -it b1ba266386d7 bash
root@kind-control-plane:/# curl http://10.96.230.243:8080
Hello World!root@kind-control-plane:/# exit

$ curl http://localhost:8080
curl: (56) Recv failure: Connection reset by peer

$ docker inspect -f "{{ .NetworkSettings.IPAddress }}" b1ba266386d7
192.168.64.2

$ curl http://192.168.64.2:8080
curl: (7) Failed to connect to 192.168.64.2 port 8080: Connection refused
```

**Environment:**

- kind version: (use `kind version`): v0.5.0, v0.5.1
- Kubernetes version: (use `kubectl version`): v1.15.3
- Docker version: (use `docker info`): 19.03.1
- OS (e.g. from `/etc/os-release`): Gentoo (also Ubuntu 16.04)

**Comment by BenTheElder** [comment]
Hi, I think there's a bit of confusion about NodePorts.

The nodeport is on `30138` in this situation, which is not forwarded. `8080` is only on the ClusterIP (`10.96.230.243`) (which is a virtual IP reachable inside the cluster).

Instead, you can use a hostPort or, even use an ingress or similar to forward from a port to in-cluster traffic.

https://kind.sigs.k8s.io/docs/user/resources/#using-contour-ingress-with-kind
https://kubernetes.io/docs/concepts/services-networking/service/#nodeport

**Comment by BenTheElder** [comment]
that is, `hello-node` is reachable at:

`10.96.230.243:8080` (in-cluster)
`192.168.64.2:30138` (`192.168.64.2` appears to be the node container IP, reachable from the host if not on macOS)

Since you are on Linux you if you want to use a NodePort you can skip the `extraPortMapping` and just use `${NODE_IP}:${NODE_PORT}` IE `92.168.64.2:30138`.

**Comment by gagara** [comment]
Thanks for explanation. Indeed 192.168.64.2:30138 works fine in my case. My bad. misunderstanding NodePort concepts. It would be nice to have in Quick start guide more detailed explanation and some e2e usecases where and how extraPortMappings feature can be used, because I still don't clearly understand how
> This can be useful if using NodePort services or daemonsets exposing host ports.

but I'm too new to k8s, so will RTFM :)

**Comment by aojea** [comment]
> > that is, `hello-node` is reachable at:
> > `10.96.230.243:8080` (in-cluster)
> > `192.168.64.2:30138` (`192.168.64.2` appears to be the node container IP, reachable from the host if not on macOS)
> 
> I'm having the same challenge, but am on macOS.
> Is there someplace explaining the limitations of the macOS version?

https://docs.docker.com/docker-for-mac/networking/#known-limitations-use-cases-and-workarounds

> Alternately, is there some recipe to get this working on macOS?

ports are mapped to localhost, in this example,

```

**Comment by BenTheElder** [comment]
we have a guide for ingress now, and are working on more detailed configuration guides.

**Comment by bgiddings** [comment]
This bug was the top result in a search for "extraportmappings", so even though it's closed, I think it's worth adding a bit more information here.

There is a guide for ingress now, but the QuickStart page still says:

> You can map extra ports from the nodes to the host machine with extraPortMappings:

With an example of mapping port 80 to 80.  It then says that this can be useful for "NodePort services or daemonsets exposing host ports."

From what I understand, port 80 would never be useful for a NodePort service because NodePort services only allocate ports on the node in the range 30000-32767.  So, while you can have a NodePort service with a "ClusterIP port" of 80, that port 80 is only available inside the cluster.

I don't know much about DaemonSets, but I imagine they could probably be set to use port 80.  But, I imagine that would probably be relatively uncommon.  Conventionally port 80 is for end user accessible HTTP services, since it's the only port a user doesn't have to explicitly specify in a web browser.

The detailed ingress guide uses extraPortMappings and shows how you can actually realistically use port 80 and 443, but it also uses the nginx ingress which includes cluster roles, config maps, a deployment, a namespace, and so-on.  The example on the quick start guide makes it seem like you might not need all that.

IMO, it would help to update the quick start documentation to show how a Mac user might use the extraPortMappings to access a NodePort service.  It could be as simple as showing a YAML snippet showing the nodePort value of say 31234 and then changing the containerPort in the example to 31234 as well.

Alternatively, the YAML from the extraPortMappings example could be removed and you could simply link to the ingress example.

tl;dr: It's still confusing to have an extraPortMappings example using containerPort 80 on the QuickStart guide.  That confusion can probably be lessened by using a valid port in the NodePort range, or by just linking to the ingress page.

**Comment by rhyek** [comment]
I cannot for the life of me get this to work. I have a similar issue here: #1294.

**Comment by BenTheElder** [comment]
> From what I understand, port 80 would never be useful for a NodePort service because NodePort services only allocate ports on the node in the range 30000-32767. So, while you can have a NodePort service with a "ClusterIP port" of 80, that port 80 is only available inside the cluster.

_technically_ the nodeport range is actually configurable, however in the default that particular port would not be useful. however there's no reason kind's mappings have to be on 80.

> I don't know much about DaemonSets, but I imagine they could probably be set to use port 80. But, I imagine that would probably be relatively uncommon. Conventionally port 80 is for end user accessible HTTP services, since it's the only port a user doesn't have to explicitly specify in a web browser.

using hostPort is an approach for "bare metal" ingress for example (e.g. ingress-nginx).

> IMO, it would help to update the quick start documentation to show how a Mac user might use the extraPortMappings to access a NodePort service. It could be as simple as showing a YAML snippet showing the nodePort value of say 31234 and then changing the containerPort in the example to 31234 as well.

PRs welcome 🙃 we have a similar guide for ingress, having one around nodeports could be great :-)
